### ISSUE TITLE
bugs in cftp when uploading shorter files over longer

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2519]
|<img alt="johnnypops's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| johnnypops reported|
|-|-|
|Trac ID|trac#2519|
|Type|defect|
|Created|2007-03-12 16:18:58Z|
|Branch|https://github.com/twisted/twisted/tree/cftp-put-over-longer-file-2519|

This fix makes the 'preserve time-stamp' option in WinSCP3 work:

```
Index: ssh/filetransfer.py
===================================================================
--- ssh/filetransfer.py (revision 19823)
+++ ssh/filetransfer.py (working copy)
@@ -851,7 +851,7 @@
 FILEXFER_ATTR_SIZE        = 0x00000001
 FILEXFER_ATTR_OWNERGROUP  = 0x00000002
 FILEXFER_ATTR_PERMISSIONS = 0x00000004
-FILEXFER_ATTR_ACMODTIME   = 0x00000009
+FILEXFER_ATTR_ACMODTIME   = 0x00000008
 FILEXFER_ATTR_EXTENDED    = 0x+80000000

 FILEXFER_TYPE_REGULAR        = 1
```
I got the protocol information from [here](http://tools.ietf.org/html/draft-ietf-secsh-filexfer-02#section-5)

This fix makes file transfer work when an **existing** file is being
over-written by a **smaller** file:

```
Index: scripts/cftp.py
===================================================================
--- scripts/cftp.py     (revision 19823)
+++ scripts/cftp.py     (working copy)
@@ -412,7 +412,8 @@
             remote = os.path.split(local)[1]
         lf = file(local, 'r')
         path = os.path.join(self.currentDirectory, remote)
-        d = self.client.openFile(path, filetransfer.FXF_WRITE|filetransfer.FXF_CREAT, {})
+        flags = filetransfer.FXF_WRITE|filetransfer.FXF_CREAT|filetransfer.FXF_TRUNC
+        d = self.client.openFile(path, flags, {})
         d.addCallback(self._cbPutOpenFile, lf)
         d.addErrback(self._ebCloseLf, lf)
         return d
@@ -442,7 +443,8 @@
         name = os.path.split(f)[1]
         remote = os.path.join(self.currentDirectory, path, name)
         log.msg((name, remote, path))
-        d = self.client.openFile(remote, filetransfer.FXF_WRITE|filetransfer.FXF_CREAT, {})
+        flags = filetransfer.FXF_WRITE|filetransfer.FXF_CREAT|filetransfer.FXF_TRUNC
+        d = self.client.openFile(remote, flags, {})
         d.addCallback(self._cbPutOpenFile, lf)
         d.addErrback(self._ebCloseLf, lf)
         d.addBoth(self._cbPutMultipleNext, files, path)
```

It took me a while to find this one :-)

<details><summary>Searchable metadata</summary>

```
trac-id__2519 2519
type__defect defect
reporter__johnnypops johnnypops
priority__highest highest
milestone__ 
branch__branches_cftp_put_over_longer_file_2519 branches/cftp-put-over-longer-file-2519
branch_author__mwh mwh
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__soc2007 soc2007
time__1173716338000000 1173716338000000
changetime__1227393807000000 1227393807000000
version__None None
owner__ 
cc__exarkun cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p commented|
|-|-|

Tagging soc2007

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

[#3010](https://github.com/twisted/twisted/issues/3010) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

Guess this isn't merged because it lacks unit tests?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Right.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_6"></a>|@mwhudson commented|
|-|-|

(In [25460]) Branching to 'cftp-put-over-longer-file-2519'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson commented|
|-|-|

(In [25461]) a failing test
refs [#2519](https://github.com/twisted/twisted/issues/2519)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_8"></a>|@mwhudson removed owner|
|-|-|

I cobbled a test together.  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @mwhudson|
|-|-|

The change in `cmd_PUT` is correctly tested, but not the one in `_cbPutMultipleNext`. Also, please name new tests 'test_foo', not 'testFoo'.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_10"></a>|@mwhudson removed owner|
|-|-|

OK, let's try that again.  It's not a *nice* test, but it is *a* test...

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @mwhudson|
|-|-|

Thanks a lot. Just one nit: `f = file(os.path.join(self.testDir, 'longerFile'),'w')` is missing a space before 'w', at both places.

Once fixed please merge!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_13"></a>|@mwhudson set status to `closed`|
|-|-|

(In [25485]) Merge cftp-put-over-longer-file-2519: fix bugs in cftp when uploading shorter files over longer

Fix two places in the conch sftp client where it failed to truncate a file that
was being overwritten by an upload.  Fix is from the reporter, tests are from
mwh.

Author: johnnypops, mwh
Reviewer: therve
Fixes: [#2519](https://github.com/twisted/twisted/issues/2519)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_14"></a>|@mwhudson set status to `reopened`|
|-|-|

(In [25486]) Revert r25485: fix bugs in cftp when uploading shorter files over longer

I can't use combinator, this was some other fix.  Sorry.

Reopens [#2519](https://github.com/twisted/twisted/issues/2519)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_15"></a>|@mwhudson set status to `closed`|
|-|-|

(In [25487]) Merge cftp-put-over-longer-file-2519: fix bugs in cftp when uploading shorter files over longer

Fix two places in the conch sftp client where it failed to truncate a file that
was being overwritten by an upload.  Fix is from the reporter, tests are from
mwh.

Author: johnnypops, mwh
Reviewer: therve
Fixes: [#2519](https://github.com/twisted/twisted/issues/2519)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
